### DISCUSSION TITLE
POD Reduced Basis Surrogate Example Fails in Parallel

### DISCUSSION THREAD

**Discussion by rfryeSigma** [idaholab/moose#27698] [Q&A Modules: General]
The [POD Reduced Basis Surrogate Example](https://mooseframework.inl.gov/modules/stochastic_tools/examples/pod_rb_surrogate.html) fails when run in parallel.

The example code in `moose/modules/stochastic_tools/examples/surrogates/pod_rb/2d_multireg` has `.i` files for a full_order model to run on 1000 samples, a trainer on 100 samples, and a surrogate emulating the full_order model on the same 1000 samples. When properly run the outputs of the full_order and the surrogate models agree to about 5 decimal digits.

The example codes cannot be run in place using the code generated by the `Makefile` in the stochastic_tools module because the `MaterialReaction` and `PODSurrogateTester` codes are not included. They are provided only in the `stochastic_tools/test` folder. I created a moose app with `stork`, modified its `Makefile` for `STOCHASTIC_TOOLS            := yes` and copied the supplementary test code to the app's `include` and `src` folders. With these modifications, I was able to run all of the `.i `files with a single processor on an M1 MAC and on an AWS Linux partition.

When I try to run the trainer with `mpiexec -n 8`, it runs through the first half of the training and reports errors in the second half, I include the output from the end of the first half through the failure:

```
sub12: 
sub12: Postprocessor Values:
sub12: +----------------+----------------+
sub12: | time           | nodal_l2       |
sub12: +----------------+----------------+
sub12: |   0.000000e+00 |   0.000000e+00 |
sub12: +----------------+----------------+
sub12: 
sub12:       0 Linear |R| = 1.501988e+03
sub12:       1 Linear |R| = 5.483291e+01
sub12:       2 Linear |R| = 2.983655e-01
sub12:       3 Linear |R| = 4.760389e-03
sub12:  Solve Converged!
sub12: 
sub12: Outlier Variable Residual Norms:
sub12:   psi: 1.501988e+03
sub12: 
sub12: Postprocessor Values:
sub12: +----------------+----------------+
sub12: | time           | nodal_l2       |
sub12: +----------------+----------------+
sub12: |   0.000000e+00 |   0.000000e+00 |
sub12: |   1.000000e+00 |   1.778131e+04 |
sub12: +----------------+----------------+
sub12: 
PODFullSolveMultiApp [sub] 13/100 samples complete!
PODFullSolveMultiApp [sub] 26/100 samples complete!
PODFullSolveMultiApp [sub] 39/100 samples complete!
PODFullSolveMultiApp [sub] 52/100 samples complete!
PODFullSolveMultiApp [sub] 64/100 samples complete!
PODFullSolveMultiApp [sub] 76/100 samples complete!
PODFullSolveMultiApp [sub] 88/100 samples complete!
PODFullSolveMultiApp [sub] 100/100 samples complete!
      Finished Executing MultiApps                                                       [  2.37 s] [  895 MB]
    Finished Executing MultiApps                                                         [  2.37 s] [  895 MB]
  Finished Executing MultiApps                                                           [  2.37 s] [  895 MB]
 Solve Skipped!


*** ERROR ***
/Users/rogerfrye/Documents/Git/multi_ai/problems/POD_trainer.i:56.3:
The following error occurred in the MultiApp 'sub' of type PODFullSolveMultiApp.

Internal error, a processor has an undefined app.

application called MPI_Abort(MPI_COMM_WORLD, 1) - process 5


*** ERROR ***
/Users/rogerfrye/Documents/Git/multi_ai/problems/POD_trainer.i:56.3:
The following error occurred in the MultiApp 'sub' of type PODFullSolveMultiApp.

Internal error, a processor has an undefined app.

application called MPI_Abort(MPI_COMM_WORLD, 1) - process 6


*** ERROR ***
/Users/rogerfrye/Documents/Git/multi_ai/problems/POD_trainer.i:56.3:
The following error occurred in the MultiApp 'sub' of type PODFullSolveMultiApp.

Internal error, a processor has an undefined app.

application called MPI_Abort(MPI_COMM_WORLD, 1) - process 7
mpiexec -n 8 ../multi_ai-opt -i POD_trainer.i  0.01s user 0.01s system 0% cpu 8.589 total

```

This error does not show up when `num_rows` in the trainer is less than 15 or for certain other values like 50 and 60. With `num_rows = 15`, it fails when run in parallel with 3 processors.

**Comment by GiudGiud** [comment]
@grmnptr @somu15
